src/v/compression/internal/gzip_compressor.cc

```diff
@@ -117,106 +117,114 @@ class gzip_decompression_codec {
+        throw_if_zstream_error(
+          "gzip error compressing chunk: {} (flush={})", deflate_r);
     }
-    iobuf ret;
+
+    obuf.trim(strm.total_out - ret.size_bytes());
     ret.append(std::move(obuf));
+
     return ret;
 }
 
-void gzip_decompression_codec::inflate_to(char* output, size_t out_size) {
-    size_t consumed_bytes = 0;
+iobuf gzip_decompression_codec::inflate_to_iobuf() {
+    // Max memory allocation
+    const size_t max_chunk_size = details::io_allocation_size::max_chunk_size;
+
+    // Rough guess at compression ratio to guess initial chunk size for
+    // small buffers.
+    size_t chunk_size = std::min(max_chunk_size, _input.size_bytes() * 3);
+
     int code = 0;
-    // NOLINTNEXTLINE
-    auto out = reinterpret_cast<unsigned char*>(output);
+    iobuf output;
     do {
-        // NOLINTNEXTLINE
-        _stream.next_out = out + consumed_bytes;
-        _stream.avail_out = out_size - consumed_bytes;
+        chunk_size = std::min(max_chunk_size, chunk_size * 2);
+
+        ss::temporary_buffer<char> tmp(chunk_size);
+        _stream.next_out = reinterpret_cast<unsigned char*>(tmp.get_write());
+        _stream.avail_out = chunk_size;
+
         code = inflate(&_stream, Z_NO_FLUSH);
         switch (code) {
         case Z_STREAM_ERROR:
         case Z_NEED_DICT:
         case Z_DATA_ERROR:
         case Z_MEM_ERROR:
-            throw_zstream_error("gzip uncmpress error:{}", code);
+            throw_zstream_error("gzip uncompress error:{}", code);
         default: /*do nothing*/;
         }
-        /* Advance output pointer (in pass 2). */
-        consumed_bytes = out_size - _stream.avail_out - consumed_bytes;
-    } while (_stream.avail_out == 0 && code != Z_STREAM_END);
-}
 
-static ss::temporary_buffer<char>
-buffer_for_input(const char* src, size_t src_size) {
-    auto codec = gzip_decompression_codec(src, src_size);
-    codec.reset();
-    std::array<char, 512> dummy_buf{};
-    // find gzip header
-    codec.inflate_to(dummy_buf.data(), dummy_buf.size());
-    return ss::temporary_buffer<char>(codec.stream().total_out);
-}
+        if (_stream.avail_in == 0) {
+            _input_chunk++;
+            if (_input_chunk != _input.end()) {
+                _stream.next_in = const_cast<unsigned char*>(
+                  reinterpret_cast<const unsigned char*>(_input_chunk->get()));
+                _stream.avail_in = _input_chunk->size();
+            }
+        }
 
-static iobuf do_uncompress(const char* src, size_t src_size) {
-    ss::temporary_buffer<char> buf;
-    {
-        auto codec = gzip_decompression_codec(src, src_size);
-        codec.reset();
-        buf = buffer_for_input(src, src_size);
-        // main data decompression
-        codec.inflate_to(buf.get_write(), buf.size());
+        size_t written_out_this_iter = chunk_size - _stream.avail_out;
+        tmp.trim(written_out_this_iter);
+        output.append(std::move(tmp));
+    } while (code == Z_OK && _stream.avail_in > 0);
```

Comment:
fun fact: i think that clang-tidy is going to start warning about `do {} while ()` loops in the future